nncf/common/quantization/quantizer_propagation/grouping.py

```diff
@@ -0,0 +1,102 @@
+from typing import Dict
+from typing import Optional
+from typing import Set
+
+from nncf.common.quantization.quantizer_propagation.structs import PropagatingQuantizer
+
+
+class UnifiedScalePropagatingQuantizerGroupManager:
+    """
+    Keeps track of the groups of quantizers that have to have their scales unified in the final
+    quantized model.
+    """
+    def __init__(self):
+        self._next_gid = 0
+        self._group_vs_prop_quants_dict = {}  # type: Dict[int, Set[PropagatingQuantizer]]
+
+    def _get_next_gid(self) -> int:
+        retval = self._next_gid
+        self._next_gid += 1
+        return retval
+
+    def register_group(self, prop_quants: Set[PropagatingQuantizer]) -> int:
+        for pq in prop_quants:
+            for gid, group in self._group_vs_prop_quants_dict.items():
+                assert pq not in group, "Propagating quantizer #{} is already registered in a group {}!".format(pq.id,
+                                                                                                                gid)
+        gid = self._get_next_gid()
+        self._group_vs_prop_quants_dict[gid] = prop_quants
+        return gid
+
+    def add_to_group(self, target_gid: int, prop_quant: PropagatingQuantizer):
+        for gid, group in self._group_vs_prop_quants_dict.items():
+            if target_gid != gid:
+                assert prop_quant not in group, "Tried to add propagating quantizer #{} to group #{}, " \
+                                                "but it is already registered in a group {}!".format(prop_quant.id,
+                                                                                                     target_gid,
+                                                                                                     gid)
+        self._group_vs_prop_quants_dict[target_gid].add(prop_quant)
+
+    def remove_from_group(self, group: int, prop_quant: PropagatingQuantizer):
+        self._group_vs_prop_quants_dict[group].remove(prop_quant)
+
+    def get_group_vs_prop_quants_dict(self) -> Dict[int, Set[PropagatingQuantizer]]:
+        return copy(self._group_vs_prop_quants_dict)
+
+    def get_group_id_by_propagating_quantizer_id(self, requested_pqid: int) -> Optional[int]:
+        for gid, group in self._group_vs_prop_quants_dict.items():
+            for pq in group:
+                if pq.id == requested_pqid:
+                    return gid
+        return None
+
+    def merge_groups(self, merge_to_gid: int, merge_from_gid: int):
+        if merge_to_gid == merge_from_gid:
+            return
+        self._group_vs_prop_quants_dict[merge_to_gid].update(self._group_vs_prop_quants_dict[merge_from_gid])
+        self._group_vs_prop_quants_dict.pop(merge_from_gid)
+
+
+class QuantizersWaitingForMergeManager:
+    """
+    Tracks the quantizers that await a merge while trying to transition through a downward-branching node
+    and corresponding node keys.
+    """
+
+    def __init__(self):
```

Comment:
Could you please add docstrings for all "public" class methods?

Reply:
Done